### DISCUSSION TITLE
! operator to include name of the null property on exception

### DISCUSSION THREAD

**Discussion by 0xor1** [dotnet/csharplang#7392] [Language Ideas] [answered]
At present we never use the ! ignore null operator because we lose the context of what was null on a null exception we instead use the extension methods:

```csharp

    public static T NotNull<T>([System.Diagnostics.CodeAnalysis.NotNull] this T? obj, [CallerArgumentExpression("obj")] string? paramName = null) where T : class
    {
      ArgumentNullException.ThrowIfNull((object) obj, paramName);
      return obj;
    }

    public static T NotNull<T>([System.Diagnostics.CodeAnalysis.NotNull] this T? obj, [CallerArgumentExpression("obj")] string? paramName = null) where T : struct
    {
      ArgumentNullException.ThrowIfNull((object) obj, paramName);
      return obj.Value;
    }

```

so that:

```csharp

We!.Dont!.Do!.ThisSyntax()

We.NotNull().Do.NotNull().Do.NotNull().ThisSyntax()
```

would it be possible for the ! operator to act in a similar way and throw the name of the null property?

**Comment by HaloFour** [comment]
The `!` null forgiveness operator intentionally does not throw any exceptions.  It is intended to be used in those cases where the developer wants to suppress a null dereference warning when they know that the expression will not be null.

**Reply by 0xor1** [reply]
> when they know that the expression will not be null

developers are wrong all the time, and in such instances the NullException that is thrown loses context, and you may not know which particular access was null, it may not be appropriate to change the ! operator but to change the nature of nullExceptions or create a new operator that behaves as the extension methods above. Just a thought based on my experiences using, and then not using the ! operator in favor of our own extension methods.

**Reply by HaloFour** [reply, marked as answer]
See: https://github.com/dotnet/runtime/issues/3858
#6034
